### ISSUE TITLE
Queries projecting entities incorrectly try to inline parts of their projection even when client side projection of those elements is required

### ISSUE THREAD

**Issue by espentveit** [aspnet/EntityFramework#6040]
### Steps to reproduce
```
var list =
	context.FlexLmLicenseFiles
	.Select(
	 a => new FeatureVM
	 {
		 ExpirationDate = a.Expires,
		 FeatureName = a.Feature,
		 LicenseFile = a.Service.LicenseFilePath,
		 LicenseServer = a.Service.Server.PrincipalName,
		 Location = a.Service.Server.PrincipalName,
		 Port = a.Service.PortNumber,
		 Id = a.Id,
		 Vendor = a.Vendor,
		 Version = a.Version,
		 Capacity = a.Capacity.OrderByDescending(b => b.DateTime).Select(b => b.Value).FirstOrDefault(),
		 Used = a.Usage.OrderByDescending(b => b.DateTime).Select(b => b.Value).FirstOrDefault()
	 }
	).ToList();
```
### The issue

The system should generate a query that selects the data within an acceptable time frame.

Tested with Linq to SQL (EF6). Gives a working query:
```
SELECT [t0].[Id], [t0].[Feature] AS [FeatureName], (
    SELECT TOP (1) [t3].[Value]
    FROM [FlexLmCapacities] AS [t3]
    WHERE [t3].[LicenseFileId] = [t0].[Id]
    ORDER BY [t3].[DateTime] DESC
    ) AS [Capacity], (
    SELECT TOP (1) [t4].[Value]
    FROM [FlexLmUsages] AS [t4]
    WHERE [t4].[LicenseFileId] = [t0].[Id]
    ORDER BY [t4].[DateTime] DESC
    ) AS [Used], [t0].[Expires] AS [ExpirationDate], [t0].[Vendor], [t2].[PrincipalName] AS [LicenseServer], [t1].[LicenseFilePath] AS [LicenseFile], [t1].[PortNumber] AS [Port], [t0].[Version]
FROM [FlexLmLicenseFiles] AS [t0]
LEFT OUTER JOIN [FlexLmServices] AS [t1] ON [t1].[Id] = [t0].[ServiceId]
LEFT OUTER JOIN [FlexLmServers] AS [t2] ON [t2].[Id] = [t1].[ServerId]

```

Tested with EF Core:
Output:
```
warn: Microsoft.EntityFrameworkCore.Query.Internal.SqlServerQueryCompilationContextFactory[8]
      The LINQ expression 'DefaultIfEmpty()' could not be translated and will be evaluated locally. To configure this warning use the DbContextOptionsBuilder.ConfigureWarnings API (event id 'RelationalEventId.QueryClientEvaluationWarning'). ConfigureWarnings can be used when overriding the DbContext.OnConfiguring method or using AddDbContext on the application service provider.
warn: Microsoft.EntityFrameworkCore.Query.Internal.SqlServerQueryCompilationContextFactory[8]
      The LINQ expression 'DefaultIfEmpty()' could not be translated and will be evaluated locally. To configure this warning use the DbContextOptionsBuilder.ConfigureWarnings API (event id 'RelationalEventId.QueryClientEvaluationWarning'). ConfigureWarnings can be used when overriding the DbContext.OnConfiguring method or using AddDbContext on the application service provider.
warn: Microsoft.EntityFrameworkCore.Query.Internal.SqlServerQueryCompilationContextFactory[8]
      The LINQ expression 'from FlexLmService a.Service in {[a.Service_group] => DefaultIfEmpty()}' could not be translated and will be evaluated locally. To configure this warning use the DbContextOptionsBuilder.ConfigureWarnings API (event id 'RelationalEventId.QueryClientEvaluationWarning'). ConfigureWarnings can be used when overriding the DbContext.OnConfiguring method or using AddDbContext on the application service provider.
warn: Microsoft.EntityFrameworkCore.Query.Internal.SqlServerQueryCompilationContextFactory[8]
      The LINQ expression 'join FlexLmServer a.Service.Server in value(Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1[GeoCloud.Data.Models.License.FlexLmServer]) on IIF(([a.Service] != null), Property([a.Service], "ServerId"), null) equals Convert(Property([a.Service.Server], "Id"))' could not be translated and will be evaluated locally. To configure this warning use the DbContextOptionsBuilder.ConfigureWarnings API (event id 'RelationalEventId.QueryClientEvaluationWarning'). ConfigureWarnings can be used when overriding the DbContext.OnConfiguring method or using AddDbContext on the application service provider.
warn: Microsoft.EntityFrameworkCore.Query.Internal.SqlServerQueryCompilationContextFactory[8]
      The LINQ expression 'DefaultIfEmpty()' could not be translated and will be evaluated locally. To configure this warning use the DbContextOptionsBuilder.ConfigureWarnings API (event id 'RelationalEventId.QueryClientEvaluationWarning'). ConfigureWarnings can be used when overriding the DbContext.OnConfiguring method or using AddDbContext on the application service provider.
warn: Microsoft.EntityFrameworkCore.Query.Internal.SqlServerQueryCompilationContextFactory[8]
      The LINQ expression 'DefaultIfEmpty()' could not be translated and will be evaluated locally. To configure this warning use the DbContextOptionsBuilder.ConfigureWarnings API (event id 'RelationalEventId.QueryClientEvaluationWarning'). ConfigureWarnings can be used when overriding the DbContext.OnConfiguring method or using AddDbContext on the application service provider.
warn: Microsoft.EntityFrameworkCore.Query.Internal.SqlServerQueryCompilationContextFactory[8]
      The LINQ expression 'from FlexLmServer a.Service.Server in {[a.Service.Server_group] => DefaultIfEmpty()}' could not be translated and will be evaluated locally. To configure this warning use the DbContextOptionsBuilder.ConfigureWarnings API (event id 'RelationalEventId.QueryClientEvaluationWarning'). ConfigureWarnings can be used when overriding the DbContext.OnConfiguring method or using AddDbContext on the application service provider.
warn: Microsoft.EntityFrameworkCore.Query.Internal.SqlServerQueryCompilationContextFactory[8]
      The LINQ expression '(Convert(Property([a], "Id")) == Property([b], "LicenseFileId"))' could not be translated and will be evaluated locally. To configure this warning use the DbContextOptionsBuilder.ConfigureWarnings API (event id 'RelationalEventId.QueryClientEvaluationWarning'). ConfigureWarnings can be used when overriding the DbContext.OnConfiguring method or using AddDbContext on the application service provider.
warn: Microsoft.EntityFrameworkCore.Query.Internal.SqlServerQueryCompilationContextFactory[8]
      The LINQ expression 'FirstOrDefault()' could not be translated and will be evaluated locally. To configure this warning use the DbContextOptionsBuilder.ConfigureWarnings API (event id 'RelationalEventId.QueryClientEvaluationWarning'). ConfigureWarnings can be used when overriding the DbContext.OnConfiguring method or using AddDbContext on the application service provider.
warn: Microsoft.EntityFrameworkCore.Query.Internal.SqlServerQueryCompilationContextFactory[8]
      The LINQ expression '(Convert(Property([a], "Id")) == Property([b], "LicenseFileId"))' could not be translated and will be evaluated locally. To configure this warning use the DbContextOptionsBuilder.ConfigureWarnings API (event id 'RelationalEventId.QueryClientEvaluationWarning'). ConfigureWarnings can be used when overriding the DbContext.OnConfiguring method or using AddDbContext on the application service provider.
warn: Microsoft.EntityFrameworkCore.Query.Internal.SqlServerQueryCompilationContextFactory[8]
      The LINQ expression 'FirstOrDefault()' could not be translated and will be evaluated locally. To configure this warning use the DbContextOptionsBuilder.ConfigureWarnings API (event id 'RelationalEventId.QueryClientEvaluationWarning'). ConfigureWarnings can be used when overriding the DbContext.OnConfiguring method or using AddDbContext on the application service provider.

info: Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommandBuilderFactory[1]
      Executed DbCommand (0ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT [a.Service.Server].[Id], [a.Service.Server].[AssetStatus], [a.Service.Server].[CustomerId], [a.Service.Server].[DisplayName], [a.Service.Server].[Notes], [a.Service.Server].[ObjectStatus], [a.Service.Server].[Path], [a.Service.Server].[PrincipalName]
      FROM [FlexLmServers] AS [a.Service.Server]

info: Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommandBuilderFactory[1]
      Executed DbCommand (253ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT [a].[Id], [a].[AssetStatus], [a].[DisplayName], [a].[Expires], [a].[Feature], [a].[FeatureTxt], [a].[Licenses], [a].[Notes], [a].[ObjectStatus], [a].[Path], [a].[PrincipalName], [a].[ServiceId], [a].[ServiceName], [a].[Vendor], [a].[Version], [a.Service].[Id], [a.Service].[AssetStatus], [a.Service].[DisplayName], [a.Service].[LicenseFilePath], [a.Service].[LogFilePath], [a.Service].[Notes], [a.Service].[ObjectStatus], [a.Service].[Path], [a.Service].[PortNumber], [a.Service].[PrincipalName], [a.Service].[ProgramFile], [a.Service].[ServerId], [a.Service].[ServiceName], [a.Service].[StartMode], (
          SELECT TOP(1) [b].[Value]
          FROM [FlexLmCapacities] AS [b]
          WHERE [a].[Id] = [b].[LicenseFileId]
          ORDER BY [b].[DateTime] DESC
      ), (
          SELECT TOP(1) [b0].[Value]
          FROM [FlexLmUsages] AS [b0]
          WHERE [a].[Id] = [b0].[LicenseFileId]
          ORDER BY [b0].[DateTime] DESC
      )
      FROM [FlexLmLicenseFiles] AS [a]
      LEFT JOIN [FlexLmServices] AS [a.Service] ON [a].[ServiceId] = [a.Service].[Id]
      ORDER BY [a].[ServiceId]
```

For each row in the result Entity sends the following queries. Which makes the query take a very long time:
```
info: Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommandBuilderFactory[1]
      Executed DbCommand (9ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT [b1].[LicenseFileId], [b1].[Value]
      FROM [FlexLmCapacities] AS [b1]
      ORDER BY [b1].[DateTime] DESC
info: Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommandBuilderFactory[1]
      Executed DbCommand (9ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT [b2].[LicenseFileId], [b2].[Value]
      FROM [FlexLmUsages] AS [b2]
      ORDER BY [b2].[DateTime] DESC

```

### Further technical details

EF Core version: 1.0.0. Also tested with v1.1.0-alpha1-21573
Operating system: Windows 10
Visual Studio version: VS 2015

Other details about my project setup:
Models:
```
public class FlexLmCapacity
{
	public int Id { get; set; }
	public DateTime DateTime { get; set; }
	public int Value { get; set; }
	public FlexLmLicenseFile LicenseFile { get; set; }
}

public class FlexLmLicenseFile
{
	public int Id { get; set; }
	public string Path { get; set; }
	public string Feature { get; set; }
	public string FeatureTxt { get; set; }
	public string Version { get; set; }
	public string Licenses { get; set; }
	public string Expires { get; set; }
	public string Vendor { get; set; }
	public string ObjectStatus { get; set; }
	public string AssetStatus { get; set; }
	public string Notes { get; set; }
	public string DisplayName { get; set; }
	public string ServiceName { get; set; }
	public string PrincipalName { get; set; }
	public FlexLmService Service { get; set; }
	public List<FlexLmUsage> Usage { get; set; }
	public List<FlexLmCapacity> Capacity { get; set; }
}

public class FlexLmServer
{
	public int Id { get; set; }
	public string Path { get; set; }
	public string ObjectStatus { get; set; }
	public string AssetStatus { get; set; }
	public string Notes { get; set; }
	public string DisplayName { get; set; }
	public string PrincipalName { get; set; }
	public List<FlexLmService> Services { get; set; }
}

public class FlexLmService
{
	public int Id { get; set; }
	public string Path { get; set; }
	public string ServiceName { get; set; }
	public string StartMode { get; set; }
	public string ProgramFile { get; set; }
	public string LicenseFilePath { get; set; }
	public string LogFilePath { get; set; }
	public string PortNumber { get; set; }
	public string ObjectStatus { get; set; }
	public string AssetStatus { get; set; }
	public string Notes { get; set; }
	public string DisplayName { get; set; }
	public string PrincipalName { get; set; }
	public List<FlexLmLicenseFile> LicenseFiles { get; set; }
	public FlexLmServer Server { get; set; }
}

public class FlexLmUsage
{
	public int Id { get; set; }
	public DateTime DateTime { get; set; }
	public int Value { get; set; }
	public FlexLmLicenseFile LicenseFile { get; set; }
}

public class FeatureVM : IEntity
{
	public int Id { get; set; }
	public string FeatureName { get; set; }
	public int? Capacity { get; set; }
	public int? Used { get; set; }
	public int? Available { get; set; }
	public string Expired { get; set; }
	public string ExpirationDate { get; set; }
	public string Vendor { get; set; }
	public string LicenseServer { get; set; }
	public string LicenseFile { get; set; }
	public string Port { get; set; }
	public string Version { get; set; }
	public string Location { get; set; }
	public string Status { get; set; }
}
```

**Comment by maumar** [comment]
duplicate of https://github.com/aspnet/EntityFramework/issues/4588

**Comment by maumar** [comment]
@espentveit this has also been improved significantly by https://github.com/aspnet/EntityFramework/commit/05b24c30b8b8693bc3daa78b195831bf45749a3f

We still produce multiple subqueries, but the filters are now translated so SQL, so the queries look like so:

```
SELECT TOP(1) [b1].[Value]
FROM [Capacities] AS [b1]
WHERE @_outer_Id = [b1].[LicenseFileId]
ORDER BY [b1].[DateTime] DESC',N'@_outer_Id int',@_outer_Id=1
```

**Comment by Sjark** [comment]
@maumar why is there created queries for each result in the example above?

As far as I can see, the two first queries created by EF should be the only queries needed to get the correct result back, so the query per result should not be needed at all.

The following is the sql generated by EF Core:
```
info: Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommandBuilderFactory[1]
      Executed DbCommand (0ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT [a.Service.Server].[Id], [a.Service.Server].[AssetStatus], [a.Service.Server].[CustomerId], [a.Service.Server].[DisplayName], [a.Service.Server].[Notes], [a.Service.Server].[ObjectStatus], [a.Service.Server].[Path], [a.Service.Server].[PrincipalName]
      FROM [FlexLmServers] AS [a.Service.Server]

info: Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommandBuilderFactory[1]
      Executed DbCommand (253ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT [a].[Id], [a].[AssetStatus], [a].[DisplayName], [a].[Expires], [a].[Feature], [a].[FeatureTxt], [a].[Licenses], [a].[Notes], [a].[ObjectStatus], [a].[Path], [a].[PrincipalName], [a].[ServiceId], [a].[ServiceName], [a].[Vendor], [a].[Version], [a.Service].[Id], [a.Service].[AssetStatus], [a.Service].[DisplayName], [a.Service].[LicenseFilePath], [a.Service].[LogFilePath], [a.Service].[Notes], [a.Service].[ObjectStatus], [a.Service].[Path], [a.Service].[PortNumber], [a.Service].[PrincipalName], [a.Service].[ProgramFile], [a.Service].[ServerId], [a.Service].[ServiceName], [a.Service].[StartMode], (
          SELECT TOP(1) [b].[Value]
          FROM [FlexLmCapacities] AS [b]
          WHERE [a].[Id] = [b].[LicenseFileId]
          ORDER BY [b].[DateTime] DESC
      ), (
          SELECT TOP(1) [b0].[Value]
          FROM [FlexLmUsages] AS [b0]
          WHERE [a].[Id] = [b0].[LicenseFileId]
          ORDER BY [b0].[DateTime] DESC
      )
      FROM [FlexLmLicenseFiles] AS [a]
      LEFT JOIN [FlexLmServices] AS [a.Service] ON [a].[ServiceId] = [a.Service].[Id]
      ORDER BY [a].[ServiceId]
```

That should be all that is needed to generate the `List<FeatureVM>`

**Comment by maumar** [comment]
@Sjark that is a good point, we shouldn't be doing this twice. Will re-open and investigate

**Comment by maumar** [comment]
@espentveit currently you can make things better by splitting the query into two:

```cs
var list1 =
    context.FlexLmLicenseFiles
    .Select(
     a => new 
     {
         ExpirationDate = a.Expires,
         FeatureName = a.Feature,
         Id = a.Id,
         Vendor = a.Vendor,
         Version = a.Version,
         Capacity = a.Capacity.OrderByDescending(b => b.DateTime).Select(b => b.Value).FirstOrDefault(),
         Used = a.Usage.OrderByDescending(b => b.DateTime).Select(b => b.Value).FirstOrDefault()
     }
    ).ToList();

var list2 =
    context.FlexLmLicenseFiles
    .Select(
     a => new 
     {
         LicenseFile = a.Service.LicenseFilePath,
         LicenseServer = a.Service.Server.PrincipalName,
         Location = a.Service.Server.PrincipalName,
         Port = a.Service.PortNumber,
         Id = a.Id,
     }
    ).ToList();

// combine list1 and list2 on the client
```

**Comment by maumar** [comment]
fixed in f2632eb0fb9104427042625fb81ab8e135f3bb2c

**Comment by Sjark** [comment]
I have now tested the fix, and it is not generating the sub queries now (intended?).

The query now takes about 15 seconds with the current fix, in EF6 it only sends one query to the database that has to sub queries, and gets the result back in under 1 second.

At the moment, our FlexLmLicenseFiles table got 472 rows, so with the current query plan, the above linq example will send 472 * 2 queries to the DB server, instead of 1 query with 2 sub queries.

Is this something you are looking at fixing for the 1.1 release?

**Comment by maumar** [comment]
@Sjark lack of subqueries is intended. The fix is pretty much about removing those subqueries - they were not being used in the end result (we don't know how to flow it to the result objects). 

The problem should go away once #4588 is fixed and it's scheduled for 1.1.0

In the meantime you can use the workaround provided above - splitting the query into two parts, one containing fields accessed through optional navigations and one containing the rest.
